File: backend/app/schemas/external_sources.py
"""Schemas for external provider status metadata."""

from __future__ import annotations

from enum import Enum

from pydantic import BaseModel


class ExternalSourceState(str, Enum):
    """Operational state for an external provider capability."""

    LIVE = "live"
    MOCK = "mock"
    UNAVAILABLE = "unavailable"


class ExternalSourceMetadata(BaseModel):
    """Machine-readable metadata for external provider provenance."""

    provider: str
    state: ExternalSourceState
    configured: bool
    synthetic: bool
    reason: str | None = None

File: frontend/src/api/externalSources.ts
import { boolish, coerceString } from './shared'

export type ExternalSourceState = 'live' | 'mock' | 'unavailable'

export interface ExternalSourceMetadata {
  provider: string
  state: ExternalSourceState
  configured: boolean
  synthetic: boolean
  reason?: string | null
}

export function mapExternalSourceMetadata(
  payload: unknown,
): ExternalSourceMetadata | null {
  if (!payload || typeof payload !== 'object') {
    return null
  }

  const source = payload as Record<string, unknown>
  const provider = coerceString(source.provider)
  const rawState = coerceString(source.state)?.toLowerCase()
  if (!provider || !rawState) {
    return null
  }

  const state: ExternalSourceState =
    rawState === 'live' || rawState === 'mock' || rawState === 'unavailable'
      ? rawState
      : 'unavailable'

  return {
    provider,
    state,
    configured: boolish(source.configured),
    synthetic: boolish(source.synthetic),
    reason: coerceString(source.reason) ?? null,
  }
}
